ncm-network/src/main/perl/nmstate.pm

```diff
@@ -0,0 +1,720 @@
+# return array of hashref
+sub make_nm_ip_route
+{
+    my ($self, $device, $routes, $routing_table_hash) = @_;
+    my @rt_entry;
+    foreach my $route (@$routes) {
+        my %rt;
+        if ($route->{address} eq 'default') {
+                $rt{destination} = '0.0.0.0/0';
+        } else {
+             if ($route->{netmask}){
+                 my $dest_addr = NetAddr::IP->new($route->{address}."/".$route->{netmask});
+                 $rt{destination} = $dest_addr->cidr;
+             } else {
+                # if no netmask defined for a route, assume its single ip
+                $rt{destination} = $route->{address}."/32";
+             }
+        }
+        $rt{'table-id'} = "$routing_table_hash->{$route->{table}}" if $route->{table};
+        $rt{'next-hop-interface'} = $device;
+        $rt{'next-hop-address'} = $route->{gateway} if $route->{gateway};
+        push (@rt_entry, \%rt);
+
+    }
+    return \@rt_entry;
+}
+
+# group all eth bound to a bond together in a hashref for to be used as
+# - port in nmstate config file
+sub get_bonded_eth
+{
+    my ($self, $interfaces) = @_;
+    my @data =  ();
+    foreach my $name (sort keys %$interfaces) {
+        my $iface = $interfaces->{$name};
+        if ( $iface->{master} ){
+            push @data, $name;
+        }
+    }
+    return \@data;
+}
+
+# writes the nmstate yml file, using yaml module.
+sub nmstate_file_dump
+{
+    my ($self, $filename, $ifaceconfig) = @_;
+    # ATM interfaces hash will only have one entry per interface. so looking at first entry is fine. long as file isn't resolv.yml
+    my $iface = $ifaceconfig->{'interfaces'}[0] if ($filename ne $NM_RESOLV_YML);
+
+    my $changes = 0;
+
+    my $func = "nmstate_file_dump";
+    my $testcfg = $self->testcfg_filename($filename);
+    if (! defined($self->cleanup($testcfg, undef, keeps_state => 1))) {
+        $self->warn("Failed to cleanup testcfg $testcfg before file_dump: $self->{fail}");
+    }
+
+    if (!$self->file_exists($filename) || $self->mk_bu($filename, $testcfg))
+    {
+        my $trd = EDG::WP4::CCM::TextRender->new('yaml', $ifaceconfig, relpath => 'network');
+        if (! defined($trd->get_text()))
+        {
+            $self->error ("Unable to generate network config $filename: $trd->{fail}");
+            return;
+        };
+        my $fh = $trd->filewriter($testcfg,
+                                header => "# File generated by " . __PACKAGE__ . ". Do not edit",
+                                log => $self);
+        my $filestatus;
+        if ($fh->close()) {
+            if ($self->file_exists($filename)) {
+                $self->info("$func: file $filename has newer version scheduled.");
+                $filestatus = $UPDATED;
+            } else {
+                $self->info("$func: new file $filename scheduled.");
+                $filestatus = $NEW;
+            }
+        } else {
+            if ($filename ne $NM_RESOLV_YML)
```

Comment:
invert the logc and swap the blocks: instead of 
```
if not something
   x
else
  y
```
it's easier to read and understand to do
```
if something:
  y
else
  x
```